doc/api18970.md

```diff
@@ -0,0 +1,288 @@
+
+* Kernel [initialization
+  level](https://docs.zephyrproject.org/latest/reference/drivers/index.html#initialization-levels)
+* User-space versus kernel (system call)
+* Normal ("thread"?) versus interrupt (invoked from an Interrupt Service
+  Routine)
+
+For API behavior we are interested in whether a particular function
+**may**, **must**, or **must not** be invoked from a specific context.
+
+### Function vs Operation
+
+tl;dr: A function *returns* while an operation *completes*.
+
+A **function** is an addressible sequence of actions to which control is
+transferred by **invoking** the function with various parameters, and
+from which control is transferred back to the point where it was invoked
+when the function **returns**.
+
+A function may succeed or fail.  In many cases failure is indicated by a
+negative integer return value while success is indicated by a
+non-negative integer return value, but specific functions may use other
+conventions.
+
+An **operation** is behavior, which is generally initiated by invoking a
+function.  An operation can also succeed or fail.  An operation
+**completes** when the success or failure of the operation has been
+determined and made available through whatever mechanism was specified
+through the initiating function.
+
+For APIs we are particularly interested in the relationship between the
+function and its associated operation.  Some potential relationships
+include:
+* The function succeeds if and only if its associated operation
+  completes with success. (This is usually expected of *synchronous*
+  functions.)
+* If the function fails then its associated operation has not been
+  initiated.  If the function succeeds then its associated operation has
+  been initiated.  The success or failure of the operation is not
+  otherwise coupled to success or failure of its initiating function.
+  (This is usually expected of *asynchronous* functions.)
+
+## Definitions
+
+### suspendable (function)
+
+A function is suspendable when invocation by a cooperative thread may
+trigger a context switch.
+
+#### Commentary
+
+For thread safety it is generally necessary to hold a lock of some sort.
+If a sequence of code involves only invocation of non-suspendable
+functions then the lock may be avoidable.
+
+See also discussion of the term "block" at the definition of
+*synchronous*.
+
+Note that if a suspendable function is marked as isr-callable then it
+must internally check `k_is_in_isr()` and behave as a no-wait function,
+whether or not it was invoked in a no-wait mode.
+
+### no-wait (function)
+
+A suspendable function is no-wait when calls to it can be forced by a
+parameter into a no-wait mode, i.e. one that takes a path that will not
+trigger a context switch from a cooperative thread.
+
+#### Commentary
+
+A no-wait function invoked in no-wait mode in conditions where it cannot
+initiate its operation should fail, in many cases by returning the value
+`-EWOULDBLOCK`.
+
+### isr-callable (function)
```

Comment:
I wonder if you want to generalize this notion to encompass being callable from what the Linux kernel calls "atomic context".

Focusing on ISRs might be a bit short-sighted -- what if we're not in an ISR, but do hold a spinlock in thread context on SMP? Would we need "spinlock-callable" too?

Reply:
That was the thought behind the original question about whether we needed a `k_is_atomic()`.  For now we have only `k_is_in_isr()` and I want two things: (1), a clear definition of what something must promise in order to be unconditionally callable from an ISR (in the latest version that's *reschedule-safe*), and (2) a way to denote that a function allows a parameter to ensure it's callable even when its behavior would normally not be safe (that'd be *no-wait*).

Reply:
> That was the thought behind the original question about whether we needed a `k_is_atomic()`.

I don't follow that logic, sorry.

The relevant point in the LWN article I keep linking to is that functions can be safe (or not) to call from atomic context, entirely independently of whether or not there exists a predicate that returns true if and only if the current code is running in atomic context.

You seem to be talking about the latter (existence of a predicate). I am talking about the former (whether a code is OK to call from atomic context as a generalization of interrupt context).

What am I missing?

Reply:
The short version is that I'm focusing on terms that I think I understand.  I don't know what to do about atomic.

Can you explain what you mean by "atomic context"?  If the cases we're trying to address are:
* can be safely invoked from an interrupt handler; versus
* can be safely invoked when a spinlock is held

is there really a relationship between them?

Maybe the thing to do is to start with defining the conditions for something being "atomic-callable".  If you agree, please propose something: I have no idea what that would mean that isn't covered by "reschedule-safe".

Reply:
Resolved offline, I was seeing things that weren't in the document. Sorry for noise.